### ISSUE TITLE
sessionDestroyed is not being called in latest DW version

### ISSUE THREAD

**Issue by orangecoding** [dropwizard/dropwizard#2031]
We've upgraded the dw version and noticed that the sessionDestroyed in our internal event handler is not being called anymore.

I've reported this to jetty in the first place, they've tracked it down and finally noticed you are implementing a wrong Interface.
Please see the issue I'Ve reported in the jetty repo here: https://github.com/eclipse/jetty.project/issues/1432#issuecomment-297665805

> So all dropwizard needs to do to make this work is to change RoutingHandler to respect the HandlerContainer interface.

I've also a testcase for you.
http://www.orange-coding.net/dw_test.zip

build it with `mvn clean install -DskipTests`, afterwards go to ./target start it with `java -jar Test-1.2.0-SNAPSHOT.jar server` then point your browser to localhost:8080.

In the System.out, you'll see a CREATED (Session has been created) but no DESTROY. (Should appear after 10 seconds)

**Comment by janbartel** [comment]
If the dropwizard folks need any further help, the jetty team will be happy to answer any questions.

**Comment by orangecoding** [comment]
👍 Thanks!

**Comment by t-tsutsumi** [comment]
I also saw the same thing. Probably, the `SessionHandler` will work with [this](https://github.com/dropwizard/dropwizard/compare/master...t-tsutsumi:pr/implement-HandlerContainer). However, there is concern about maintaining backward compatibility. I think more tests will be required.

Also, If you run the test case of `@orangecoding`, I recommend adding the following code.
```diff
diff --git a/src/main/java/com/example/test/TestSessionHandler.java b/src/main/java/com/example/test/TestSessionHandler.java
index 75726cd..5d7924b 100644
--- a/src/main/java/com/example/test/TestSessionHandler.java
+++ b/src/main/java/com/example/test/TestSessionHandler.java
@@ -3,6 +3,10 @@ package com.example.test;
 import javax.servlet.http.HttpSessionEvent;
 import javax.servlet.http.HttpSessionListener;

+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.SessionIdManager;
+import org.eclipse.jetty.server.session.DefaultSessionIdManager;
+import org.eclipse.jetty.server.session.HouseKeeper;
 import org.eclipse.jetty.server.session.SessionHandler;

 public class TestSessionHandler extends SessionHandler{
@@ -24,4 +28,18 @@ public class TestSessionHandler extends SessionHandler{
     });

   }
+
+  // Sets the interval of the scavenge task to 5 seconds.
+  @Override
+  protected void doStart() throws Exception {
+    Server server = getServer();
+    HouseKeeper houseKeeper = new HouseKeeper();
+    houseKeeper.setIntervalSec(5);
+    SessionIdManager sessionIdManager = new DefaultSessionIdManager(server);
+    sessionIdManager.setSessionHouseKeeper(houseKeeper);
+    server.setSessionIdManager(sessionIdManager);
+    server.manage(sessionIdManager);
+    sessionIdManager.start();
+    super.doStart();
+  }
 }

```

**Comment by nickbabcock** [comment]
@janbartel, I tried swapping `AbstractHandler` for `AbstractHandlerContainer`, but I can still repro @orangecoding's bug. Am I missing something? I looked at [HotSwapHandler](https://github.com/eclipse/jetty.project/blob/0c8273f2ca1f9bf2064cd9c4c939d2546443f759/jetty-server/src/main/java/org/eclipse/jetty/server/handler/HotSwapHandler.java) for a bit of guidance. I've included the diff done to our own `RoutingHandler`

```diff
diff --git a/dropwizard-jetty/src/main/java/io/dropwizard/jetty/RoutingHandler.java b/dropwizard-jetty/src/main/java/io/dropwizard/jetty/RoutingHandler.java
index 4281ec6..6d7461b 100644
--- a/dropwizard-jetty/src/main/java/io/dropwizard/jetty/RoutingHandler.java
+++ b/dropwizard-jetty/src/main/java/io/dropwizard/jetty/RoutingHandler.java
@@ -3,15 +3,17 @@ package io.dropwizard.jetty;
 import org.eclipse.jetty.server.Connector;
 import org.eclipse.jetty.server.Handler;
 import org.eclipse.jetty.server.Request;
-import org.eclipse.jetty.server.handler.AbstractHandler;
+import org.eclipse.jetty.server.handler.AbstractHandlerContainer;
 
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
 
-public class RoutingHandler extends AbstractHandler {
+public class RoutingHandler extends AbstractHandlerContainer {
     /**
      * We use an array of entries instead of a map here for performance reasons. We're only ever
      * comparing connectors by reference, not by equality, so avoiding the overhead of a map is
@@ -29,6 +31,7 @@ public class RoutingHandler extends AbstractHandler {
     }
 
     private final Entry[] entries;
+    private final Handler[] handlers;
 
     public RoutingHandler(Map<Connector, Handler> handlers) {
         this.entries = new Entry[handlers.size()];
@@ -37,6 +40,9 @@ public class RoutingHandler extends AbstractHandler {
             this.entries[i++] = new Entry(entry.getKey(), entry.getValue());
             addBean(entry.getValue());
         }
+
+        final HashSet<Handler> handleSet = new HashSet<>(handlers.values());
+        this.handlers = handleSet.toArray(new Handler[handleSet.size()]);
     }
 
     @Override
@@ -53,5 +59,17 @@ public class RoutingHandler extends AbstractHandler {
             }
         }
     }
+
+    @Override
+    public Handler[] getHandlers() {
+        return handlers.clone();
+    }
+
+    @Override
+    protected void expandChildren(List<Handler> list, Class<?> byClass) {
+        for (Handler handler : handlers) {
+            expandHandler(handler, list, byClass);
+        }
+    }
 }
 

```

**Comment by nickbabcock** [comment]
@t-tsutsumi thanks for the tip! That is what I was missing. I guess I didn't want to wait the default 11mins for a scavenge.

**Comment by janbartel** [comment]
@nickbabcock  not sure why your code isn't working, it looks approximately right to me.  @t-tsutsumi does your code work?

**Comment by nickbabcock** [comment]
@janbartel sorry, it does work! I just didn't wait 11minutes. After I set the house keeper to 5 seconds, it's working.

**Comment by janbartel** [comment]
@nickbabcock  excellent. And it would be great to have a way to change the scavenge interval via dropwizard setup.

**Comment by orangecoding** [comment]
well, maybe this is a stupid question, but if I set `setMaxInactiveInterval` to a certain number of seconds, I would expect the session to be destroyed after that time. At least it was like that before I've updated. Am I wrong? 

This is what the docu says.
> Sets the max period of inactivity, after which the session is invalidated, in seconds.

**Comment by t-tsutsumi** [comment]
@nickbabcock OK, I will submit a pull request.

@orangecoding I think that the session will be destroyed by **the next access** after the `maxInactiveInterval` has elapsed, or the scavenge task.

**Comment by nickbabcock** [comment]
Closed by #2032
